wgsl/index.bs

```diff
@@ -1312,6 +1312,18 @@ The spelling of the token may be the same as an [=identifier=], but the token do
 
 Section [[#context-dependent-name-tokens]] lists all such tokens.
 
+## Built-in Value Names ## {#builtin-value-names}
```

Comment:
It feels incongruous with built-in value names and interpolation controls separate from the other context dependent names. I'd prefer some sort of unification.

I can see two main ways to accomplish this:
1. Move these to the context dependent name tokens section.
2. Pull the current context dependent name tokens section here. All the various types of context dependent names could be subsections. The main section would commonize the lack of resolution. Subsections could be brief descriptions and grammar tables.

I think 2 is more in line with with this PR and your stated reasoning in the description.

Reply:
I totally agree- there are further merits to option two, such as reducing the exhaustive re-enumeration of attribute names when they are quite explicit in spec.

In case @dneto0 is also not feeling against going the way 2, and if we think this change will not bloat the PR, I can go ahead and implement it here. Otherwise, can delegate to subsequent PR, too. Guidance would be appreciated, thank you!

Reply:
Thank you very much once again. This PR no longer contains any changes for those sections, as normative PR took care of it. Hence, I mark this as resolved; any subsequent improvements can probably be separate PRs or issues. Thank you for your understanding in advance.

Reply:
I think I'd still prefer the reorganization done as part of this editorial PR. So my preference would be the following:

3.8 Context-dependent Names
3.8.1 Attribute Names
3.8.2 Diagnostic Severity Control
3.8.3 Diagnostic Rule Names
3.8.4 Extension Names
3.8.5 Swizzle Names
3.8.6 Built-in Value Names
3.8.7 Interpolation Type Names
3.8.8 Interpolation Sampling Names

Then remove 15.4 Context-Dependent Name Tokens (or make just link back to 3.8). Thoughts?

Reply:
I think the direction is good (including simply removal to reduce long jump in spec), and I can certainly do this reorganization as part of this PR, too, I'm just neutral about it. If @dneto0 agrees with the direction and implementing here, then I'll go forward and push a commit. Thank you!

Reply:
The way the PR is right now,    `3.9 Built-in Value Names` has these parts:
* (a) Its section name
* (b) The definition saying a built-in value name is a token naming built-in values. 
* (c) The rule about keywords, reserved words, and not resolving as an identifer
* (d) A reference to the section of the spec that describes usage. "See 12.3.1.1 Built-in Inputs and Outputs".
* (e) A grammar rule expanding to ident_pattern_token

My observations:
* (b) is essential
* (d) is redundant because (b) links to the concept already, and that's in the section that (d) refers to.
* (c) is common, as Alan pointed out. I think we modify the definition (b) to say it's a context-dependent name, and then allow the resolution rules for context-dependent names to take over.  So we don't need to say (c) at all.
* (e) is almost content-free.


It feels pretty heavyweight to have a whole section for what's left.
Having whole sections for causes a lot of jumping around. For example:

* Let's say I'm in [12.3.1.4 Interpolation](https://gpuweb.github.io/gpuweb/wgsl/#interpolation). Its first sentence says:
*  >  Authors can control how user-defined IO data is interpolated through the use of the [interpolate](https://gpuweb.github.io/gpuweb/wgsl/#attribute-interpolate) attribute.
* Now I follow the link to the 'interpolate' attribute.  In this PR it goes to [11.8 interpolate](https://mehmetoguzderin.github.io/gpuweb-202404-attributes/index.html#attribute-interpolate) which is a whole section to give the grammar rule and describe the parameters.  
* In 11.8 the grammar rule in links to `interpolate_type_name` which jumps  back to  the grammar rule given in 3.11.
* In 11.8 the description in of the values links to `interpolation type name-token` which jumps to that definition, also in 3.11
* Now, `3.11 Interpolation Type Names` doesn't actually list the names. For that you have to follow `interpolation type` *back to 12.3.1.4 Interpolation*.  

So to know what you can write for an interpolation attribute you end up jumping to three different sections and end up back where we started.  That's a bad reader experience.

I think that jumping around is inevitable if we make new sections for each token kind.

I would rather pull the contents of 15.4 Context-dependent Name Tokens directly into 3.8, so 3.8 would give the general description *and* all the actual lists.  Note that the *current* spec has 3.9 .. 3.12 that I think should be collapsed away.  (That is, the current spec has the jumping-around problem).

Reply:
@dneto0 In the sentence

> Having whole sections for causes a lot of jumping around. For example

Between "for" and "causes" words, did you mean having whole sections for name-tokens, such as interpolation type names, etc.? If so, I can follow the directions in your last paragraph; I also think that taking 15.4 into section three does give a better reading experience. Should I just link to section 3 or remove 15.4?

Thank you!

Reply:
Right, sorry for the missing words. Having whole sections for the specific name tokens is one problem. (like 3.9 .. 3.12).

So let's eliminate 3.9 to 3.12.

It seems bad to jump from 11.8 to 3.11.  I think one of your main goals is to improve the readability of attribute descriptions.  So you changed the table in section 11 into subsections.  That's fine, and I'm sure is much better on a narrow viewport like a phone. 

I wonder about pulling the grammar rule for `interpolate_type_name` into `11.8 interpolate`, so it sits right under the `interpolate_attr` rule.  And in the text say it's a context-dependent name. Something like.

> Valid Values	One or two parameters.

> The first parameter is a context-dependent name for one of the interpolation kinds.

> The first parameter is a _context-dependent name_. It must name one of the _interpolation types_.

> The second parameter, if present, is a context-dependent-name. It must name one of the _interpolation sampling_ modes.

If we do this then the reader only bounces between two sections: 11.8 and 12.3.1.4.

Folding 15.4 into 3 is logically distinct, but then again both moves modify section 3 as it is now.

Reply:
@dneto0 I think pulling the grammar rule into `11.8 interpolate` is the ideal version, and the step up I had in mind but decide to not go that far before submitting the PR. Then I think a commit can add all three following to improve the situation in the best way possible:

* Eliminate 3.9 to 3.12,
* Fold 15.3 into 3,
* Pull grammar rule for `interpolate_type_name` into `11.8 interpolate`

@alan-baker do you think such a commit would be good and also improve in the direction you request change in this thread?

Thank you!

Reply:
Before review of @alan-baker I took one last look and made subsections than to eliminate entirely, more closely matching https://github.com/gpuweb/gpuweb/pull/4554#discussion_r1558143738 feedback, which does navigate and read a bit better. Preview is updated: https://mehmetoguzderin.github.io/gpuweb-202404-attributes/index.html

Thank you!